Add QR code in post action ellipsis menu

Initsogar opened the pull request:
#### Proposed Changes

* Add an option to generate a QR code in `Posts` and `Pages` lists, specifically in the ellipsis menu.
* This feature will be disabled in `production` for now, since the feature is behind a feature flag. It will be available on local and staging environments.

#### Screenshots
Posts list | Pages list | QR Modal
---|---|---
<img width="222" alt="image" src="https://user-images.githubusercontent.com/1488641/199628036-4a1858b9-7a39-4352-afca-3aa0211e2547.png"> | <img width="211" alt="image" src="https://user-images.githubusercontent.com/1488641/199628250-1aff4e93-68cf-4d06-bade-f301701c9555.png"> | <img width="257" alt="image" src="https://user-images.githubusercontent.com/1488641/199628197-9ee37f99-a65c-4445-b1ae-606e47ce57f4.png">

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Clone the repository and checkout to this branch, see `GET STARTED` on the README section of this repo.
2. Start Calypso blue `yarn start`
3. Navigate to `http://calypso.localhost:3000/`
4. Switch to your site and navigate to `Posts` section.
5. Click on the ellipsis menu (the tree dots) on a published post, and you should see the new QR Code option.
6. Click on `QR Code` button and a modal dialog will be displayed with the QR code of the post selected.
7. Check with your phone if the QR code is pointing to the expected URL.
8. Do step 5, 6 and 7 again on `Pages` section.

**Testing feature enablement**
This feature will not be available in production yet, so we need to ensure that this feature is only enable where we config `post-list/qr-code-link` to `true`. For now, the feature will be available on the development and staging environments (proxied). 
* Make sure you see the QR code option in the post list dropdown.
* Go to your development.json file and change the value of `post-list/qr-code-link` to false.
* Ctrl+C your `yarn start` command and restart it.
* Make sure you don't see the QR code option in the post list dropdown, since the feature flag is turned off.

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] ~[Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?~
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] ~Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)~
- [ ] ~Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?~

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Initsogar committed:
Add QR code in post action ellipsis menu

spsiddarthan committed:
Added the flag post-list/qr-code-link.

Initsogar committed:
Add QR dialog for posts list

spsiddarthan committed:
Add QR Code to page.

spsiddarthan committed:
Move position down.

spsiddarthan committed:
Remove changes to jetpack cloud config files.

spsiddarthan committed:
Remove changes to jetpack cloud config files.

spsiddarthan committed:
enable feature on staging.

Initsogar committed:
Add QR svg icon for the new QR code option

Initsogar committed:
Merge branch 'add/qr-post-list' of https://github.com/Automattic/wp-calypso into add/qr-post-list

spsiddarthan committed:
update flag.

spsiddarthan committed:
Merge branch 'add/qr-post-list' of https://github.com/Automattic/wp-calypso into add/qr-post-list

naxoc commented on client/my-sites/pages/page/index.js:
```diff
@@ -351,6 +352,13 @@ class Page extends Component {
+		if ( ! config.isEnabled( 'post-list/qr-code-link' ) ) {
+			return null;
+		}
+		return <PostActionsEllipsisMenuQRCode globalId={ this.props.page.global_ID } key="bug" />;
```
Is the 'bug' key still good here now that we have our own icon?

Initsogar replied:
Done!

naxoc commented on client/my-sites/post-actions-qr-code/index.jsx:
```diff
@@ -0,0 +1,30 @@
+import QRCode from 'qrcode.react';
+import { useCallback, useState, useEffect } from 'react';
+
+function PostActionsQRCode( { siteUrl, showQRCode } ) {
```
Can we add prop-types?

Initsogar replied:
Done!

naxoc commented on client/my-sites/post-type-list/post-actions-ellipsis-menu/qrcode.jsx:
```diff
@@ -0,0 +1,73 @@
+import { Icon } from '@wordpress/components';
+import { localize } from 'i18n-calypso';
+import PropTypes from 'prop-types';
+import { Component } from 'react';
```
I _think_ that we can import this from the `@wordpress` package thing.

Initsogar replied:
Done!

Initsogar committed:
Add PropTypes to PostActionsQRCode component

Initsogar committed:
Update key for PostActionsEllipsisMenuQRCode

Initsogar committed:
Use Component from wordpress/element instead react

Initsogar committed:
Merge branch 'add/qr-post-list' of https://github.com/Automattic/wp-calypso into add/qr-post-list

Initsogar committed:
Update key for PostActionsEllipsisMenuQRCode

naxoc reviewed (approved):
Nice work! This looks good to me now :shipit:

spsiddarthan committed:
Merge branch 'trunk' of https://github.com/Automattic/wp-calypso into add/qr-post-list

tyxla commented on client/my-sites/post-actions-qr-code/index.jsx:
```diff
@@ -0,0 +1,36 @@
+	}, [] );
+
+	return (
+		<>
```
Is the fragment necessary? Since we have a single child, it seems like we can remove it.

Initsogar replied:
Yeah, shouldn't be there. It was added while debugging some stuff. We will remove it.

tyxla commented on config/horizon.json:
```diff
@@ -80,6 +80,7 @@
 		"plugins/ssr-categories": false,
 		"plugins/ssr-landing": true,
 		"post-editor/checkout-overlay": true,
+		"post-list/qr-code-link": false,
```
A false flag won't be necessary. They're all disabled by default, and have to be specifically enabled for every environment.

Initsogar replied:
I was thinking about it, but then I thought: wouldn't it be better to set it as false as a way of saying "hey, we introduced a new feature and if you wanted here, you need to set it true!". We are not including it (even on false) on jetpack-cloud environments because that feature actually doesn't work there, but in case of Calypso Blue I think it would be okay to declare it at least somehow. What do you think?

tyxla replied:
I see your point of view, but I really see all those false flags as unnecessary noise. If they have no real purpose, then they should be removed IMHO. I don't feel so strongly about it, though, so I'll leave it to your judgement.

tyxla commented on config/production.json:
```diff
@@ -95,6 +95,7 @@
 		"plugins/ssr-categories": false,
 		"plugins/ssr-landing": true,
 		"post-editor/checkout-overlay": true,
+		"post-list/qr-code-link": false,
```
Same as above.

tyxla commented on config/test.json:
```diff
@@ -71,6 +71,7 @@
 		"plans/starter-plan": false,
 		"plugins/ssr-categories": true,
 		"plugins/ssr-landing": true,
+		"post-list/qr-code-link": false,
```
Same as above.

tyxla commented on client/my-sites/post-type-list/post-actions-ellipsis-menu/qrcode.jsx:
```diff
@@ -0,0 +1,73 @@
+import { getPost } from 'calypso/state/posts/selectors';
+import PostActionsQRCode from '../../post-actions-qr-code';
+
+class PostActionsEllipsisMenuQRCode extends Component {
```
This looks like a great candidate for a functional component. Any particular reason to make it a class component instead?

Initsogar replied:
Timely reasons, mostly. We were trying to use one of the existing post action ellipsis menu components as a reference to create this one. But yeah, I agree we should refactor it into a functional component, I think we can do it in the next iteration.

tyxla replied:
> we can do it in the next iteration

Sounds good to me. Those are famous last words, though 😅

Initsogar replied:
We iterated it in the same PR, so we are okay for now :D

tyxla commented on client/my-sites/post-type-list/post-actions-ellipsis-menu/qrcode.jsx:
```diff
@@ -0,0 +1,73 @@
+
+class PostActionsEllipsisMenuQRCode extends Component {
+	static propTypes = {
+		globalId: PropTypes.string,
```
Seems like this component won't do much without a `globalId`. Should this be a required prop?

Initsogar replied:
It seems that specifying it as a required prop is throwing a console error. Probably because it is trying to render all menu options on loading placeholder and there's no `globalId` at that moment. I will leave it as optional for now (as we do currently on the others menu items) and will revisit this in another PR.

tyxla commented on client/my-sites/post-type-list/post-actions-ellipsis-menu/qrcode.jsx:
```diff
@@ -0,0 +1,73 @@
+		const { translate, url, status } = this.props;
+
+		// The QR option will be available only for `publish` status
+		if ( ! [ 'publish' ].includes( status ) ) {
```
If we have a single item here, we might as well make it an equality comparison:

```suggestion
		if ( status !== 'publish' ) {
```

tyxla commented on client/my-sites/post-type-list/post-actions-ellipsis-menu/qrcode.jsx:
```diff
@@ -0,0 +1,73 @@
+		status: PropTypes.string,
+	};
+
+	constructor( props ) {
```
In case we decided to stay with a class component, the `constructor()` is unnecessary. We could inline the default state directly in the component class:

```
state = {
	showQRCode: false,
};
```

tyxla commented on client/my-sites/post-type-list/post-actions-ellipsis-menu/qrcode.jsx:
```diff
@@ -0,0 +1,73 @@
+					icon={
+						<Icon
+							icon={
+								<svg className="gridicon needs-offset-y" viewBox="0 0 24 24" width="18" height="18">
```
This svg could be declared outside of the component.

tyxla commented on client/my-sites/post-type-list/post-actions-ellipsis-menu/qrcode.jsx:
```diff
@@ -0,0 +1,73 @@
+		this.setState( { showQRCode: true } );
+	};
+
+	render() {
```
Should we bail early if there's no `url` or `status`?

tyxla reviewed (approved):
This is looking good from the frameworks perspective 👍 

I've left a few suggestions, nothing serious or blocking.

Nice work ✨

Initsogar committed:
Refactor PostActionsEllipsisMenuQRCode

Initsogar committed:
Remove unnecessary fragment

Initsogar committed:
Refactor PostActionsEllipsisMenuQRCode component

Moving from a class component to a functional component.

naxoc commented on client/my-sites/post-type-list/post-actions-ellipsis-menu/qrcode.jsx:
```diff
@@ -0,0 +1,61 @@
+
+	return (
+		<>
+			<PopoverMenuItem onClick={ generateQRCode } icon={ <Icon icon={ qrIcon } /> }>
```
Small thing, but might help readability: Would it be possible to wrap the `<Icon>` part already in the `qrIcon` const?

Initsogar replied:
Agree. I'm on it!

Initsogar replied:
Done

Initsogar committed:
Refactor qrIcon

tyxla commented on client/my-sites/post-type-list/post-actions-ellipsis-menu/qrcode.jsx:
```diff
@@ -0,0 +1,65 @@
+	globalId: PropTypes.string.isRequired,
+};
+
+const mapStateToProps = ( state, { globalId } ) => {
```
Using `connect()` is not something we generally do with functional components. Instead, it's preferred that we use the `useSelector()` hook directly in the functional component. That should also help simplify the code a bit.

Initsogar replied:
Thanks for the recommendation! You are right. Just refactored the component to use `useSelector()`.

tyxla commented on client/my-sites/post-type-list/post-actions-ellipsis-menu/qrcode.jsx:
```diff
@@ -0,0 +1,65 @@
+	};
+};
+
+export default connect( mapStateToProps )( localize( PostActionsEllipsisMenuQRCode ) );
```
Just as with `connect()`, the `localize()` HoC is not recommended for use in functional components. Instead, we use the `useTranslate()` hook directly in the component.

Initsogar replied:
Now we are using `useTranslate()`. Thanks for the recommendation!

tyxla commented on client/my-sites/post-type-list/post-actions-ellipsis-menu/qrcode.jsx:
```diff
@@ -0,0 +1,65 @@
+			<svg className="gridicon needs-offset-y" viewBox="0 0 24 24" width="18" height="18">
+				<path
+					d="M1,1h1v1h-1z M2,1h1v1h-1z M3,1h1v1h-1z M4,1h1v1h-1z M5,1h1v1h-1z M6,1h1v1h-1z M7,1h1v1h-1z M9,1h1v1h-1z M11,1h1v1h-1z M12,1h1v1h-1z M13,1h1v1h-1z M15,1h1v1h-1z M16,1h1v1h-1z M17,1h1v1h-1z M18,1h1v1h-1z M19,1h1v1h-1z M20,1h1v1h-1z M21,1h1v1h-1z M1,2h1v1h-1z M7,2h1v1h-1z M12,2h1v1h-1z M13,2h1v1h-1z M15,2h1v1h-1z M21,2h1v1h-1z M1,3h1v1h-1z M3,3h1v1h-1z M4,3h1v1h-1z M5,3h1v1h-1z M7,3h1v1h-1z  ...
+					fill-color="#000000"
```
Nit: is there a chance this might inadvertently take precedence over the color scheme of the app? Should we find a different place to style it, one that's more compatible with the color scheming capabilities of Calypso?

Initsogar replied:
Just removed the fill-color prop. This icon is using the same properties as `gridicon` so it is using the same color scheme.

We are expecting to introduce a new qr-code icon on gridicon soon, but we think this is okay as a workaround for now.

Initsogar committed:
Replace connect() and mapStateToProps()

The idea is to use useSelector() and useTranslate() instead of connect + localize.

tyxla reviewed (approved):
Nothing further from my perspective. Nice work 🚀

Initsogar committed:
Merge remote-tracking branch 'origin/trunk' into add/qr-post-list
